### PR TITLE
Bootstrap rewrite

### PR DESCRIPTION
**Pull request by memnac**
@timantrobus - added 3 x files as part of revision:
- bootstrap.sh (consolidated script)
- client.conf (external config file)
- cleanup-bootstrap.sh (optional - can scrap).

Script fails on step 3-1 deploying bootstrap resources. Logs report error on "(gcloud.builds.submit) unable to read file > cloudbuild.yaml".

Kindly test script in your environment. The final two steps can be independently tested by specifying 6 & 7 as second arguments.

### CODE REVIEW

File: bootstrap.sh [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    echothis "$(timestamp) [4-3]: Migrating terraform state from local to GCS remote..."
+    terraform init -migrate-state -force-copy | tee -a $LOG_FILE_PATH >&3
+    echothis $STEP_SUCCESS; sleep 1
+
+# >>>> STEP 5 - Generate github keys and configure git global <<<<
+    echothis "\n>>>>>>>>>> RUNNING: STEP #5.\n\n"
+
+    echothis "$(timestamp) [5-1]: Generate keys to connect to private Github repo..."
+    ssh-keygen -t rsa -b 4096 -N "" -q -C "${GITHUB_BOT_USER}" -f ~/.ssh/id_github 2>&1> /dev/null # suppress output
+    ssh-keyscan -t rsa github.com 2>&1 | tee ~/.ssh/known_hosts > /dev/null 2&>1 # suppress output
```
Comment (timantrobus) [inline bootstrap.sh:217]: Would be good to add a check here to confirm if these ssh key files are already present. Otherwise if present, the scripts fails

**Review by timantrobus — CHANGES_REQUESTED** [review]
Hi @memnac - mostly comments here around usability improvements. Under testing the script runs well and provides a consistent and clear structure to the bootstrap process.

The visibility around progress of the deployment and reporting errors to the console will all help with the real world application of this script.

File: bootstrap.sh [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    echothis $STEP_SUCCESS; sleep 1
+
+    echothis "$(timestamp) [2-4]: Generate terraform plan..."
+    if [ -f "${PWD}/$TF_PLAN" ]
+    then
+        echothis "Found existing plan! Replacing..."
+        rm $TF_PLAN
+    fi
+    
+    terraform plan -out=$TF_PLAN
```
[the remark this answers is not in the document]
Comment (memnac) [reply bootstrap.sh]: Amended script as advised and expanded to cover other commands as well. 
New commands will need to have this code snippet appended at the end :
`" | tee -a $LOG_FILE_PATH >&3; check_exit_code $?"`
This snippet ensures the command output is written to log file and STDOUT. The exit status code of command is also checked to handle errors gracefully.

File: bootstrap.sh [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    echothis "$(timestamp) [2-5]: Instruction - Review the generated plan before deploying. Run 'terraform show bootstrap.tfplan'.\n"; sleep 2
+
+    pause "[CHECKPOINT #2]: Do you want to continue? [Y/N]: " # CHECKPOINT!!!!
+
+# >>>> STEP 3 - Deploy bootstrap resources <<<<<<<
+    echothis "\n>>>>>>>>>> RUNNING: STEP #3.\n\n"
+    echothis "$(timestamp) [3-1]: Deploying the bootstrap infrastructure..."
+    
+    terraform apply -auto-approve $TF_PLAN
+
```
[the remark this answers is not in the document]
Comment (memnac) [reply bootstrap.sh:242]: Script amended as advised.

File: bootstrap.sh [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+# REUSABLE FUNCTION TO PREFIX STDOUT MESSAGES/COMMENTS WITH TIMESTAMP
+function timestamp()
+{
+    date +"%Y-%m-%d %T"
+}
+
+# >>>> STEP 0 - Source conf file and create log file if missing <<<<<<<
+printf "\n>>>>>>>>>> RUNNING: STEP #0.\n\n"
+printf "$(timestamp) [0-0]: Sourcing environment variables from .conf file..." 
+source $1 # Source config file
```
[the remark this answers is not in the document]
Comment (memnac) [reply bootstrap.sh]: Script amended as advised. Checks are included in new function block "step0_prep_tasks".

File: bootstrap.sh [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    echothis "$(timestamp) [3-2]: Instruction - Review terraform output by running 'terraform show output'.\n"; sleep 2
+    pause "[CHECKPOINT #3]: Do you want to continue? [Y/N]: " # CHECKPOINT!!!!
+
+# >>>> STEP 4 - Update and migrate terraform state to GCS backend <<<<
+    echothis "\n>>>>>>>>>> RUNNING: STEP #4.\n\n"    
+    echothis "$(timestamp) [4-1]: Updating terraform gcs backend conf with state bucket name..."
+
+    export BUCKET_NAME=$(terraform output -raw tf-state-bucket-name)
+    #export BUCKET_NAME="appdev-320214-terraform"
+
```
[the remark this answers is not in the document]
Comment (memnac) [reply bootstrap.sh:251]: Removed now. Commented out line was used for unit testing.

File: bootstrap.sh [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        esac
+    done
+    echothis "\n"
+fi
+}
+
+# START OF SWITCH CASE STATEMENTS - INTENDED FOR FOCUSED UNIT TESTING.
+case $2 in
+1) 
+# >>>> STEP 1 - Initialize variables and generate terraform.tfvars <<<<<<<
```
[the remark this answers is not in the document]
Comment (memnac) [reply bootstrap.sh]: Script amended as advised. Refactored switch statement to include an "all" and individual steps "1-7" options.
Checkpoints are enabled for the "all" option but disabled for individual steps.

File: bootstrap.sh (hunk 1 above)
Comment (memnac) [reply bootstrap.sh]: Scripts amended as advised. Checks included in function "step5_generate_sshkeys_and_configure_git".

**Review by timantrobus — APPROVED** [review]

**Pull request merged**